design/raft.md

```diff
@@ -0,0 +1,240 @@
+gets taken, logs that predate the snapshot are no longer necessary, because the
+snapshot captures all the information that's needed from the log up to that
+point. The number of old snapshots and WALs to retain is configurable.
+
+In SwarmKit's usage, WALs mostly contain protobuf-serialized data store
+modifications. A log entry can contain a batch of creations, updates, and
+deletions of objects from the data store. Some log entries contain other kinds
+of metadata, like node additions or removals. Snapshots contain a complete dump
+of the store, as well as any metadata from the log entries that needs to be
+preserved. The saved metadata includes the Raft term and index, a list of nodes
+in the cluster, and a list of nodes that have been removed from the cluster.
+
+WALs and snapshots are both stored encrypted, even if the autolock feature is
+disabled. With autolock turned off, the data encryption key is stored on disk in
+plaintext, in a header inside the TLS key. When autolock is turned on, the data
+encryption key is encrypted with a key encryption key.
+
+## Initializing a Raft cluster
+
+The first manager of a cluster (`swarm init`) assigns itself a random Raft ID.
+It creates a new WAL with its own Raft identity stored in the metadata field.
+The metadata field is the only part of the WAL that differs between nodes. By
+storing information such as the local Raft ID, it's easy to restore this
+node-specific information after a restart. In principle it could be stored in a
+separate file, but embedding it inside the WAL is most convenient.
+
+The node then starts the Raft state machine. From this point, it's a fully
+functional single-node Raft instance. Writes to the data store actually go
+through Raft, though this is a trivial case because reaching consensus doesn't
+involve communicating with any other nodes. The `Run` loop sees these writes and
+serializes them to disk as requested by the `etcd/raft` package.
+
+## Adding and removing nodes
+
+New nodes can join an existing Raft consensus group by invoking the `Join` RPC
+on the leader node. This corresponds to joining a swarm with a manager-level
+token, or promoting a worker node to a manager. If successful, `Join` returns a
+Raft ID for the new node and a list of other members of the consensus group.
+
+On the leader side, `Join` tries to append a configuration change entry to the
+Raft log, and waits until that entry becomes committted.
+
+A new node creates an empty Raft log with its own node information in the
+metadata field. Then it starts the state machine. By running the Raft consensus
+protocol, the leader will discover that the new node doesn't have any entries in
+its log, and will synchronize these entries to the new node through some
+combination of sending snapshots and log entries. It can take a little while for
+a new node to become a functional member of the consensus group, because it
+needs to receive this data first.
+
```

Comment:
Would it be useful to say here that as a node is caught up, when it is catching up by means of log entries, it basically plays each log entry, so if something is watching for changes as it is playing through those entries, they'll get lots of update events instead of one final state?

This detail caused some possible issues when I was doing CA rotation, and could be a gotcha to keep in mind.